### ISSUE TITLE
Improve KeyVault load times

### ISSUE THREAD

**Issue by AlexFlat** [aspnet/Extensions#943]
We are using the KeyVault provider and it works as expected, however our appSettings is quite large (250 keys) at the moment.
With this number of keys, the load takes approx 2mins to complete.
We are looking to reduce these keys to a minimum but would like to improve the load times anyway.

Can the 'foreach (var secretItem in secrets)' be made parallel?
Extensions/src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs
This could be an optional parameter to AddAzureKeyVault() that allows the calls to be made in parallel.

I understand this requirement is quite specific, however having an optional argument could be a good addition.

Thanks in advance,

Alex

**Comment by rynowak** [comment]
> With this number of keys, the load takes approx 2mins to complete.

If it's this slow, it's definitely worth looking into. Do you have any perf data you can share that will help us?

**Comment by davidfowl** [comment]
There was another place this was filed and I believe @pakrym is looking into it.

**Comment by AlexFlat** [comment]
> > With this number of keys, the load takes approx 2mins to complete.
> 
> If it's this slow, it's definitely worth looking into. Do you have any perf data you can share that will help us?

In my debug I am able to load 274 secrets in 108,000ms
This is from a local workstation hitting Azure KeyVault from outside the datacentre (ie not from an Azure VM).
Using these numbers, each secret takes approx 400ms to retrieve.
Are these the numbers you were after?
![image](https://user-images.githubusercontent.com/17284248/51098219-3becbe80-181d-11e9-99d3-ec1f1e2415a2.png)

**Comment by gfoidl** [comment]
> There was another place this was filed

https://github.com/aspnet/Extensions/issues/910

**Comment by pakrym** [comment]
I'll close this as a duplicate of #910.

**Comment by pakrym** [comment]
@AlexFlat have you been observing long startup time before or did it change recently?

**Comment by AlexFlat** [comment]
> @AlexFlat have you been observing long startup time before or did it change recently?

We have only started using the library so not sure if this has changed.
If there is a plan to improve the loading, can error handling be added to the loop?
Currently if there is a transient fault when retrieving any key, the entire load will fail.
For large settings this could cause multiple loads.
Can some type of transient fault handling be added at this layer, or is this logic up to the caller?
